walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,124 @@
+    */
+    private Collection<TimeRange> updateWithOptionalAttendees(Collection<TimeRange> workingTimes, 
+                                                                Collection<Event> events, 
+                                                                MeetingRequest request) {
+        boolean foundGuest = false;
+        for(Event occasion: events) {
+            for(String guest: request.getOptionalAttendees()) {
+                if(occasion.getAttendees().contains(guest)) {
+                    foundGuest = true;
+                    break;
+                }
+            }
+            if(foundGuest) {
+                int start = occasion.getWhen().start();
+                int end = occasion.getWhen().end();
+                boolean isAtEndOfDay = (end == TimeRange.END_OF_DAY)? true: false;
+                TimeRange guestMeeting = TimeRange.fromStartEnd(start, end, isAtEndOfDay);
+                for(TimeRange currentTimeRange: workingTimes) {
+                    if(currentTimeRange.contains(guestMeeting)){
+                        workingTimes = splitTimeRange(workingTimes, currentTimeRange, 
+                                                        start, end, request.getDuration());
+                        break;
+                    }
+                }
+                foundGuest = true;
+            }
+        }
+        return workingTimes;
+    }
+
+    /** Handles the cases where a new time block may be open before or after the meeting. 
+     *  Determines if the current TimeRange should be removed and update workingTimes.
+     */
+    private Collection<TimeRange> splitTimeRange(Collection<TimeRange> workingTimes, 
+                                                TimeRange candidate, int start, 
+                                                int end, long duration) {
+        int newStartTime = candidate.start();
+        int newEndTime = candidate.end();
+        boolean remove = false;
+
+        // Check block of time before meeting
+        if(newStartTime != start && meetingFits(duration, newStartTime, start)) {
+            workingTimes.add(TimeRange.fromStartEnd(newStartTime, start, false));
+            remove = true;
+        }
+        
+        // Check block of time after meeting
+        if(newEndTime != end && meetingFits(duration, end, newEndTime)) {
+            workingTimes.add(TimeRange.fromStartEnd(end, newEndTime, false));
+            remove = true;
+        }
+
+        // Check block of time during meeting
+        if(squeezedIn(start, newStartTime, end, newEndTime)) {
+            remove = true;
+        }
+        if(meetingFits(duration, start, end)) {
+            remove = true;
+        }
+        if(remove) {
+            workingTimes.remove(candidate);
+        }
+        return workingTimes;
+    }
+
+    /** Determines if it completely covers the TimeRange block */
+    private boolean squeezedIn(int start, int newStartTime, int end, int newEndTime) {
```

Comment:
Great use of helper functions in this class. It makes the code very easy to read. Small nit, prefer using active verbs in your method names to make it semantically more descriptive. For boolean methods like this one, prefixing with "is" is common. E.g. `isSqueezedIn()`